Assignment 1

luxiya01 opened the pull request:
The solutions to assignment 1 can be found in the folder called "assignment1". 
The solution to Q1 is included in the file "string_permutation.py", while that to Q2 is included in the file "single_linked_list.py".

luxiya01 committed:
Add solution to Q1

luxiya01 committed:
Add implementations for Q2

luxiya01 committed:
Change file name

muzaluisa commented:
Hey,

For the first assigment with string permutations, the solution is pretty elegant, good job! (may be it would be good to handle empty strings though?) 

p.s. It would be nicer as well to include examples of the usage, so that the code for the reviewer is Plug&Play

For the second assignment: I made the list of [0,1,2] and could not find the function to return kth last element (e.g. 1th last element would be 1), seems like you are returning rather the value of the index (e.g. return 0 for k = 0) :)  

Also print class function in Python 2.7 did not work for me. I had to change the name.

luxiya01 commented:
@muzaluisa Thank you for your fast review! I will add a test class and things tonight (*^__^*). 
As for print, maybe it's time to upgrade to Python3 XD ?

luxiya01 committed:
Add a field *size* to the single_linked_list class

luxiya01 committed:
Remove redundant function (due to having misunderstood the question...)

luxiya01 committed:
Add method retrieving data contained at the kth element counting backwards

luxiya01 committed:
Add unittest for string permutation

luxiya01 committed:
Empty strings are permutations of each other

luxiya01 committed:
Add tests for invalid inputs

luxiya01 committed:
Add unittest for linked list class

N.B. Only methods get_index and get_kth_from_last have been tested.
Only valid indices have been tested

luxiya01 committed:
Add path to exec

luxiya01 commented:
@muzaluisa Thank you so much! You made me realize that I completely misunderstood Q2 T_T. I have now added the get_kth_from_last method, and added unittests for relevant methods in both string_permutation and single_linked_list.

siimphh commented on assignment1/single_linked_list.py:
```diff
@@ -0,0 +1,83 @@
+            self.head = new_node
+        else:
+            curr = self.head
+            while curr.next != None:
```
This is implementation is fine, but just as a follow-up question: could the `append(...)` be implemented as O(1) operation?

siimphh commented on assignment1/test_linked_list.py:
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/python3.5
```
Thanks for providing tests!

siimphh commented on assignment1/string_permutation.py:
```diff
@@ -0,0 +1,18 @@
+def string_permutation(str1, str2):
+    if (str.isalpha(str1) or str1 == "") and (
+            str.isalpha(str2) or str2 == ""):
+        str1 = str1.lower()
```
Just as a general comment, calling out special behavior (like this lower-casing) or special assumptions is a good practice. For example, function could come with a docstring that mentioned it is case insensitive.

siimphh commented on assignment1/string_permutation.py:
```diff
@@ -0,0 +1,18 @@
+            str.isalpha(str2) or str2 == ""):
+        str1 = str1.lower()
+        str2 = str2.lower()
+        inputs = {x: 0 for x in string.ascii_lowercase} 
```
This is fine but just as an FYI - `setdefault(...)` and `collections.defaultdict(...)` might also be considered here:

```
inputs = collections.defaultdict(lambda: 0)
for c in str1:
  inputs[c] += 1
...
```

siimphh commented on assignment1/string_permutation.py:
```diff
@@ -0,0 +1,18 @@
+            inputs[c] += 1
+        for c in str2: 
+            inputs[c] -= 1
+        for c in inputs.keys():
```
Sometimes list comprehensions can be cryptic but it might be straightforward enough to use one here:

```
return all(c == 0 for c in inputs.values())
```

siimphh commented on assignment1/string_permutation.py:
```diff
@@ -0,0 +1,18 @@
+import string 
+
+def string_permutation(str1, str2):
+    if (str.isalpha(str1) or str1 == "") and (
```
Purely a style comment, but to make function bodies easier to read, it is generally better to do all the checks up front and then proceed with the main part of the function body as top-level code, that is:

```
def string_permutation(str1, str2):
  if not ((str.isalpha(str1) or str1 == "") and (str.isalpha(str2) or str2 == "")):
    raise ValueError(...)
  str1 = str1.lower()
  ...
```

It does not always matter, but tends to replace in easier to read ans "shallower" code. Especially if multiple layers of conditionals and loops are involved.

siimphh commented on assignment1/test_string_permutation.py:
```diff
@@ -0,0 +1,34 @@
+
+class TestStringPermutations(unittest.TestCase):
+    def setUp(self):
+        self.apple1 = "Apple"
```
I get the intention of creating symbolic names for the test strings to avoid typos messing up test cases. It's a valid thing to do.

I would not bother and would use literal strings to make the test cases below easier to follow. But it really does not matter much here and is up to the taste of the one writing the code (you).

I'm mentioning this because test cases can be slightly different than the rest of the code. For example, people occasionally accept a bit more code duplication in test cases if it makes it clearer what the test case is doing. One way to think of it is - since one generally has no test cases to test the test cases themselves, it is imperative for you to be able to look a test case and immediately see that they are doing the right thing.

siimphh reviewed (commented):
Just a few minor comments and suggestions on questions to consider. Note that it is normal to disagree with code review comments sometimes. Especially with some of the less technical comments in this review.
